herbert_core/utilities/maths/noisify.m

```diff
@@ -17,40 +17,69 @@
 %           Add noise with Poisson distribution, where the mean value at
 %           a point is the value y.
 %
-%   If no input errors, e, just set e=[]
+%   >> [yout, eout] = noisify(y,e,[factor,]'maximum_value',maxval)
+%           Add noise with Gaussian distribution, calculating the standard
+%           deviation by an externally provided maximum y value.
+%           The max value is preceded by a keyword string 'maximum_value'.
+%           Typically this value will be the overall maximum value if the
+%           data is processed page by page and the maximum must be
+%           extracted before this processing.
+%           Omitting factor will use the default value 0.1.
+%
+%   If no input errors, e, just set e=[]. Note that eout will be created
+%   regardless of whether e is present, so the first overload output is
+%   probably incomplete.
+
+% deal with optional arguments
+is_poisson= false;
+fac = 0.1;
+if nargin>=2
+    p = inputParser;
+    addRequired(p,'y');   % y compulsory
+    addRequired(p,'e');   % e compulsory if any of remaining optional/parameter arguments present
+    numeric_or_string = @(x) isnumeric(x) || ischar(x);
+    addOptional(p,'dist_or_factor',fac, numeric_or_string);  % fac (numeric) or distribution ('poisson') optional, default to fac=0.1
+    addParameter(p,'maximum_value',-999,@isnumeric);  % ymax, as parameter 'maximum_value'. Default to highly visible negative value.
+    parse(p,y,e,varargin{:});
+
+    % pick up signal max value as either default or input
+    ymax = p.Results.maximum_value;
 
-if nargin==3 && ischar(varargin{1})
-    if is_stringmatchi(varargin{1},'poisson')
-        yout=zeros(size(y));
-        for i=1:numel(y)
-            yout(i)=randpoisson(abs(y(i)));
-        end
-        eout=abs(y);  % the inpuy y is the mean and variance of the Poisson distribution
+    % vary if 'poisson' or fac present
+    if isnumeric(p.Results.dist_or_factor)
+        fac = p.Results.dist_or_factor;
+    elseif p.Results.dist_or_factor == 'poisson'
+        is_poisson = true;
     else
-        error('Unrecognised random sampling distribution')
+        error('3rd argument cannot be interpreted as a Gaussian factor or legal probability distribution')
     end
+end
+
+% Use Poisson distribution and ignore other arguments
+if is_poisson
+    yout=zeros(size(y));
+    for i=1:numel(y)
+        yout(i)=randpoisson(abs(y(i)));
+    end
```

Comment:
`y` could be a large array. Can this operation be vectorized? E.g.
```matlab
yout = randpoisson(abs(y))
```

Reply:
Guessing that since it is churning out random numbers, the generator is being advanced at each call. So not vectorizable. The equivalent built-in poissrnd does not seem to have this.